src/scanner.cc

```diff
@@ -1,346 +1,228 @@
-//#include <tree_sitter/parser.h>
-#include "tree_sitter/parser.h"
-#include "tree-sitter/lib/src/lexer.h" // Lexer, Length
-// http://tree-sitter.github.io/tree-sitter/creating-parsers#external-scanners
-// https://github.com/tree-sitter/tree-sitter/issues/281
-
-#include <stack> // std::stack
-// https://en.cppreference.com/w/cpp/container/stack
-// https://www.geeksforgeeks.org/stack-in-cpp-stl/
-
-// https://stackoverflow.com/a/27351464/10440128
-//#define DEBUG 1
-#if defined(DEBUG) && DEBUG > 0
- #include <stdio.h> // fprintf
- #define DEBUG_PRINT(fmt, args...) fprintf(stderr, "scanner.cc:%d: " fmt, __LINE__, ##args)
-#else
- #define DEBUG_PRINT(fmt, args...) // dont print
-#endif
+#include <cwctype>
+#include <iostream>
+#include <stdint.h>
+#include <tree_sitter/parser.h>
+#include <vector>
+
+namespace {
+
+using std::vector;
 
 enum TokenType {
-  STRING_FRAGMENT,
-  MULTISTRING_START,
-  MULTISTRING_PART_FIXED,
-  MULTISTRING_IPOL_START,
-  MULTISTRING_END,
+  MULTSTR_START,
+  MULTSTR_END,
+  STR_START,
+  STR_END,
+  INTERPOLATION_START,
+  INTERPOLATION_END,
 };
 
-static void advance(TSLexer *lexer) {
-  lexer->advance(lexer, false);
-}
+struct Scanner {
+  vector<uint8_t> expected_percent_count;
```

Comment:
please add comment to explain why this is `vector<int>` and not `int`

is this needed to handle nested strings?

Reply:
Yes! Exactly!